[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/doxav/astromodel_proving/blob/main/analysis/05_mechanistic_decomposition.ipynb)

In [1]:
from pathlib import Path
import os, sys, subprocess


def _running_in_colab() -> bool:
    return "COLAB_RELEASE_TAG" in os.environ or "google.colab" in sys.modules


if _running_in_colab():
    repo_url = os.environ.get("ASTROMODEL_REPO_URL", "https://github.com/doxav/astromodel_proving.git")
    repo_branch = os.environ.get("ASTROMODEL_REPO_BRANCH", "main")
    project_root = Path("/content") / "astromodel_proving"
    if not project_root.exists():
        subprocess.run(["git", "clone", "--depth", "1", "--branch", repo_branch, repo_url, str(project_root)], check=True)
    os.chdir(project_root)
    subprocess.run([sys.executable, "-m", "pip", "install", "-q", "-r", "requirements.txt"], check=True)
else:
    project_root = Path(os.environ.get("ASTROMODEL_PROJECT_ROOT", ".")).resolve()
    os.chdir(project_root)

if str(project_root) not in sys.path:
    sys.path.insert(0, str(project_root))

print(f"Project root: {project_root}")

Project root: /home/xav/code/astromodel_proving


# Step 05 - Mechanistic decomposition of accepted cell ensembles

This notebook validates the full-scope Step 05 reviewer-response pipeline. It loads the accepted Step 04 cell-level ensemble, simulates hidden outputs with the canonical local astrocyte model, decomposes accepted fits into Kir/gap/leak and local/spatial mechanism summaries, adds the phenotype-description layer reused from the characterization notebook, and records conservative claim scope.

The implementation reuses only the missing measure vocabulary from `analysis/unified_astrocyte_K_buffering_characterization_EXECUTED_SMOKE.ipynb`: signed local/spatial flux modes, `dKs_activation_score`, available/recruited surface proxies, mode vectors, and provisional phenotype tags. It does not reuse that notebook's legacy Optuna DB loader or duplicate its ODE. Step 05 can identify candidate mechanism regimes or compensation manifolds, but predictive and perturbation support remains pending until Step 06.

In [2]:
import json
import os
import pandas as pd
import matplotlib.pyplot as plt

from src.step05_mechanistic_decomposition import Step05Config, run_step05_mechanistic_decomposition


def _env_int(name: str, default: int) -> int:
    return int(os.environ.get(name, str(default)))


config = Step05Config(
    max_candidates=None,
    time_points=_env_int("ASTROMODEL_STEP05_TIME_POINTS", 120),
    bootstrap_iterations=_env_int("ASTROMODEL_STEP05_BOOTSTRAP_ITERATIONS", 8),
    random_seed=13,
    min_candidates_for_multicluster=6,
    min_cells_for_multicluster=3,
    min_cells_per_cluster=2,
    proxy_failure_downgrade_fraction=0.5,
    write_outputs=True,
)
result = run_step05_mechanistic_decomposition(project_root, config)
out_dir = project_root / "outputs" / "mechanisms"
print(json.dumps(result["analysis_summary"], indent=2))


/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))


/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))


/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))


/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))
/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))


/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))
/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))


/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))


/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))
/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))


/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))
/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))


/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))


/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))


/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))


/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))


/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))


/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))
/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))


/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))


/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))


/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))


/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))


/home/xav/code/astromodel_proving/src/astro_model.py:311: ODEintWarning: Excess work done on this call (perhaps wrong Dfun type). Run with full_output = 1 to get quantitative information.
  states = odeint(_rhs_for_odeint, z0_arr, t_arr, args=(paramdict,))


{
  "step_name": "Step 05 \u2014 Mechanistic decomposition of accepted cell ensembles",
  "input_source": "outputs/cell_fits/accepted_cell_ensembles.csv",
  "config": {
    "max_candidates": null,
    "time_points": 120,
    "t_final_ms": 50000.0,
    "n_clusters": 3,
    "bootstrap_iterations": 8,
    "random_seed": 13,
    "max_representatives": 3,
    "interpolation_points": 5,
    "small_stratum_min_cells": 3,
    "min_candidates_for_multicluster": 6,
    "min_cells_for_multicluster": 3,
    "min_cells_per_cluster": 2,
    "proxy_failure_downgrade_fraction": 0.5,
    "write_outputs": true
  },
  "n_input_candidates": 2098,
  "n_candidate_sweep_rows": 12588,
  "n_windowed_mechanism_rows": 50272,
  "n_phenotype_tag_rows": 2098,
  "n_successful_sweep_simulations": 12568,
  "n_mechanism_clusters": 3,
  "n_independent_cells": 30,
  "cluster_evidence_status": "candidate_regime_screen",
  "cluster_evidence_reason": "sufficient_candidates_and_cells_for_screen",
  "headline_claim_scope": "a

## 1. Accepted ensemble inventory

In [3]:
clusters = result["mechanism_clusters"]
flux = result["accepted_fit_mechanisms"]
inventory = clusters.groupby(["region", "condition"], as_index=False).agg(
    n_cells=("file_id", "nunique"),
    n_candidates=("candidate_id", "nunique"),
    n_clusters=("mechanism_cluster", "nunique"),
)
inventory

,region,condition,n_cells,n_candidates,n_clusters
0,DH,CONTROL,7,17,2
1,DH,MFA,6,653,3
2,DH,MFA_BA,6,662,3
3,VH,CONTROL,2,17,1
4,VH,MFA,2,8,2
5,VH,MFA_BA,7,741,3


## 2. Candidate-sweep flux decomposition table

In [4]:
flux_cols = [
    "file_id", "region", "condition", "candidate_id", "sweep", "current_na", "simulation_status",
    "I_Kir_integral", "I_kgap_integral", "I_leak_integral", "gap_to_kir_integral_ratio",
    "gap_fraction", "kir_fraction", "leak_fraction", "K_o_peak", "K_o_recovery_error", "proxy_validity_class",
]
flux[flux_cols].head(12)

,file_id,region,condition,candidate_id,sweep,current_na,simulation_status,I_Kir_integral,I_kgap_integral,I_leak_integral,gap_to_kir_integral_ratio,gap_fraction,kir_fraction,leak_fraction,K_o_peak,K_o_recovery_error,proxy_validity_class
0,1_DH_1_CONTROL,DH,CONTROL,1_DH_1_CONTROL__hybrid_optuna_087,1,50,ok,8.467608,3.118348,6.187632,0.368268,0.175448,0.476415,0.348136,6.399791,0.007850,strong
1,1_DH_1_CONTROL,DH,CONTROL,1_DH_1_CONTROL__hybrid_optuna_087,2,75,ok,19.032138,3.113720,17.477022,0.163603,0.078584,0.480332,0.441084,7.229443,0.006854,strong
2,1_DH_1_CONTROL,DH,CONTROL,1_DH_1_CONTROL__hybrid_optuna_087,3,100,ok,30.884728,3.117182,30.272429,0.100930,0.048498,0.480514,0.470988,8.198968,0.005686,strong
3,1_DH_1_CONTROL,DH,CONTROL,1_DH_1_CONTROL__hybrid_optuna_087,4,125,ok,45.951591,3.133222,46.408274,0.068185,0.032811,0.481203,0.485986,9.498253,0.004137,strong
4,1_DH_1_CONTROL,DH,CONTROL,1_DH_1_CONTROL__hybrid_optuna_087,5,150,ok,52.602541,3.144059,53.490153,0.059770,0.028782,0.481546,0.489672,10.097902,0.003433,strong
5,1_DH_1_CONTROL,DH,CONTROL,1_DH_1_CONTROL__hybrid_optuna_087,6,175,ok,56.934796,3.152274,58.090819,0.055366,0.026674,0.481772,0.491554,10.497663,0.002969,strong
6,1_DH_1_CONTROL,DH,CONTROL,1_DH_1_CONTROL__hybrid_optuna_098,1,50,ok,2.641268,1.340659,0.016178,0.507582,0.335324,0.660630,0.004046,5.467726,0.516493,strong
7,1_DH_1_CONTROL,DH,CONTROL,1_DH_1_CONTROL__hybrid_optuna_098,2,75,ok,3.461206,1.537982,0.015141,0.444349,0.306717,0.690263,0.003020,5.814110,0.784424,strong
8,1_DH_1_CONTROL,DH,CONTROL,1_DH_1_CONTROL__hybrid_optuna_098,3,100,ok,4.364189,1.758374,0.013984,0.402910,0.286541,0.711180,0.002279,6.218920,1.097547,strong
9,1_DH_1_CONTROL,DH,CONTROL,1_DH_1_CONTROL__hybrid_optuna_098,4,125,ok,5.492654,2.038272,0.012514,0.371091,0.270205,0.728137,0.001659,6.761449,1.517198,strong


## 2b. Protocol and K bath audit

Step 05 must not collapse the six current protocol into repeated 100 nA bath-drive simulations. The model already knows the historical current-specific bath values; Step 05 only applies explicit per-sweep/per-current/affine/gain overrides when they are present.


In [5]:
protocol_cols = [
    "candidate_id",
    "current_na",
    "sweep",
    "K_bath_middle_used",
    "K_bath_override_mode",
    "stim_window_start_s",
    "stim_window_end_s",
    "K_o_peak",
    "K_o_recovery_error",
]
protocol_audit = flux[protocol_cols].sort_values(["candidate_id", "current_na"])
protocol_summary = flux.groupby("candidate_id", as_index=False).agg(
    n_currents=("current_na", "nunique"),
    n_unique_kbath=("K_bath_middle_used", "nunique"),
    min_kbath=("K_bath_middle_used", "min"),
    max_kbath=("K_bath_middle_used", "max"),
    override_modes=("K_bath_override_mode", lambda s: ";".join(sorted(set(map(str, s))))),
)
protocol_summary


,candidate_id,n_currents,n_unique_kbath,min_kbath,max_kbath,override_modes
0,1_DH_1_CONTROL__hybrid_optuna_087,6,6,6.4,10.5,historical_current_kbath_scaled_by_gain
1,1_DH_1_CONTROL__hybrid_optuna_098,6,6,6.4,10.5,historical_current_kbath_scaled_by_gain
2,1_DH_2_CONTROL__hybrid_optuna_066,6,6,6.4,10.5,historical_current_kbath_scaled_by_gain
3,1_DH_2_CONTROL__hybrid_optuna_068,6,6,6.4,10.5,historical_current_kbath_scaled_by_gain
4,1_DH_2_CONTROL__hybrid_optuna_072,6,6,6.4,10.5,historical_current_kbath_scaled_by_gain
...,...,...,...,...,...,...
2093,VH_7_OG_MFA_BA__hybrid_pre_04,6,6,6.4,10.5,historical_current_kbath_scaled_by_gain
2094,VH_7_OG_MFA_BA__hybrid_pre_05,6,6,6.4,10.5,historical_current_kbath_scaled_by_gain
2095,VH_7_OG_MFA_BA__hybrid_pre_06,6,6,6.4,10.5,historical_current_kbath_scaled_by_gain
2096,VH_7_OG_MFA_BA__hybrid_pre_07,6,6,6.4,10.5,historical_current_kbath_scaled_by_gain


## 3. Mechanism-space clustering

In [6]:
cluster_evidence = clusters[[
    "file_id", "candidate_id", "mechanism_cluster", "cluster_evidence_status",
    "cluster_evidence_reason", "cluster_claim_scope", "proxy_validity_status",
]].drop_duplicates()
cluster_evidence

fig, ax = plt.subplots(figsize=(6, 4))
for label, sub in clusters.groupby("mechanism_cluster"):
    ax.scatter(sub["gap_fraction_mean"], sub["kir_fraction_mean"], s=80, label=label)
for _, row in clusters.iterrows():
    ax.annotate(str(row["candidate_id"]).split("__")[-1], (row["gap_fraction_mean"], row["kir_fraction_mean"]), fontsize=8)
ax.set_xlabel("Mean gap fraction")
ax.set_ylabel("Mean Kir fraction")
evidence_status = str(clusters["cluster_evidence_status"].iloc[0]) if "cluster_evidence_status" in clusters else "unknown"
ax.set_title(f"Step 05 mechanism-space cluster view ({evidence_status})")
ax.legend(title="Cluster")
fig.tight_layout()
plt.show()


/tmp/ipykernel_263462/2102515659.py:18: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


## 4. Flux decomposition for representative candidates

In [7]:
reps = result["representatives"]
rep_flux = flux.merge(reps[["file_id", "candidate_id", "representative_rank"]], on=["file_id", "candidate_id"], how="inner")
fig, ax = plt.subplots(figsize=(7, 4))
for (rank, candidate_id), sub in rep_flux.groupby(["representative_rank", "candidate_id"]):
    sub = sub.sort_values("current_na")
    ax.plot(sub["current_na"], sub["gap_fraction"], marker="o", label=f"rep {rank} gap")
    ax.plot(sub["current_na"], sub["kir_fraction"], marker="s", linestyle="--", label=f"rep {rank} Kir")
ax.set_xlabel("Current (nA)")
ax.set_ylabel("Flux fraction")
ax.set_ylim(0, 1)
ax.set_title("Representative Kir/gap decomposition across sweeps")
ax.legend(ncol=2, fontsize=8)
fig.tight_layout()
plt.show()
reps[["representative_rank", "file_id", "condition", "mechanism_cluster", "dominant_mechanism", "claim_scope"]]

/tmp/ipykernel_263462/525178024.py:14: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


,representative_rank,file_id,condition,mechanism_cluster,dominant_mechanism,claim_scope
0,1,DH_6_MFA,MFA,M1,Mixed,candidate_mechanism_regime_pending_step06_vali...
1,2,VH_5_MFA_BA,MFA_BA,M3,Mixed,candidate_mechanism_regime_pending_step06_vali...
2,3,DH_2_MFA_BA,MFA_BA,M3,Mixed,candidate_mechanism_regime_pending_step06_vali...


## 5. Compensation-vs-separated-mode interpolation diagnostics

In [8]:
geometry = result["geometry_classification"]
geometry[["cluster_a", "cluster_b", "geometry_screen_type", "interpolation_status", "geometry_classification", "validation_status"]]


,cluster_a,cluster_b,geometry_screen_type,interpolation_status,geometry_classification,validation_status
0,M1,M2,proxy_recovery_error_screen_not_simulated,effective_space_screened,compensation_manifold,pending_step06
1,M1,M3,proxy_recovery_error_screen_not_simulated,effective_space_screened,compensation_manifold,pending_step06
2,M2,M3,proxy_recovery_error_screen_not_simulated,effective_space_screened,compensation_manifold,pending_step06


## 6. Bootstrap stability table

In [9]:
stability = result["bootstrap_cluster_stability"]
stability


,bootstrap_iteration,n_cells,n_candidates,coassignment_stability,stability_status,stability_reason
0,1,18,1982,0.957934,stable,cell_level_resampling_against_original_centers
1,2,22,2503,0.958352,stable,cell_level_resampling_against_original_centers
2,3,20,2037,0.952624,stable,cell_level_resampling_against_original_centers
3,4,19,2513,0.966921,stable,cell_level_resampling_against_original_centers
4,5,21,2004,0.957736,stable,cell_level_resampling_against_original_centers
5,6,20,2283,0.956500,stable,cell_level_resampling_against_original_centers
6,7,22,2312,0.959441,stable,cell_level_resampling_against_original_centers
7,8,22,2630,0.951467,stable,cell_level_resampling_against_original_centers


## 7. DH/VH and condition occupancy table

In [10]:
enrichment = result["region_mechanism_enrichment"]
enrichment.sort_values(["region", "condition", "mechanism_cluster"])

,region,condition,mechanism_cluster,n_candidates,n_cells,total_candidates,total_cells,candidate_fraction,small_stratum_flag,interpretation_scope
0,DH,CONTROL,M1,5,4,17,11,0.294118,False,population_level_cell_association_not_phenotyp...
1,DH,CONTROL,M2,12,7,17,11,0.705882,False,population_level_cell_association_not_phenotyp...
2,DH,MFA,M1,357,6,653,18,0.546708,False,population_level_cell_association_not_phenotyp...
3,DH,MFA,M2,276,6,653,18,0.422665,False,population_level_cell_association_not_phenotyp...
4,DH,MFA,M3,20,6,653,18,0.030628,False,population_level_cell_association_not_phenotyp...
5,DH,MFA_BA,M1,472,6,662,18,0.712991,False,population_level_cell_association_not_phenotyp...
6,DH,MFA_BA,M2,139,6,662,18,0.209970,False,population_level_cell_association_not_phenotyp...
7,DH,MFA_BA,M3,51,6,662,18,0.077039,False,population_level_cell_association_not_phenotyp...
8,VH,CONTROL,M1,17,2,17,2,1.000000,True,population_level_cell_association_not_phenotyp...
9,VH,MFA,M1,2,1,8,2,0.250000,True,population_level_cell_association_not_phenotyp...


## 8. Windowed phenotype-description layer

The table below is descriptive and provisional. It uses accepted Step 04 ensembles and the canonical simulator, then adds the characterization notebook's missing local/spatial and recruitment measures. These tags do not become phenotype claims unless Step 06 supports prediction and perturbation robustness.

In [11]:
measure_registry = result["measure_registry_status"]
windowed = result["accepted_fit_mechanisms_windowed"]
phenotypes = result["buffering_phenotype_tags"]
mode_vectors = result["M_mode_vector_by_configuration"]

display(measure_registry)
display(phenotypes.head(20))
display(mode_vectors.head(12))

,measure_family,measure,status,rationale
0,state reconstruction,"Vm, K_o, DK_a, DK_t, K_s, currents",primary,Hidden-state reconstruction is retained for me...
1,signed flux budget,"local load, spatial export, bath source/sink",primary,Signed local/spatial budgets distinguish local...
2,mechanistic mode,D_F and D_I_elec with finite dominance margin,primary,Continuous local/spatial axes are primary; lab...
3,recruitment,dKs_activation_score,primary,Current-weighted recruitment is more mechanist...
4,effective gap mapping,P_gap_eff = d * pk,primary,Raw d and pk remain structurally confounded; t...
5,available surface,alpha2 = gamma_s_eff,primary,Interpreted as available spatial transfer surf...
6,recruited surface,alpha2 * dKs_activation_score,primary,Combines available capacity with dynamic recru...
7,isopotentiality proxy,1 / (1 + r_model),secondary,Reduced-model analogue because the ODE has no ...
8,pump proxy,K_o recovery and undershoot summaries,secondary,"Observable recovery descriptors, not literal N..."
9,deprecated,raw d as anatomical syncytium size,deprecated,"Superseded by P_gap_eff, available surface, an..."


,file_id,region,condition,candidate_id,buffering_phenotype,phenotype_window_scope,phenotype_specificity_score,dominant_signed_flux_mode,dKs_activation_score_mean,long_range_distribution_fraction_mean,voltage_coupling_score_mean,kir_current_score_mean,phenotype_claim_scope
0,1_DH_1_CONTROL,DH,CONTROL,1_DH_1_CONTROL__hybrid_optuna_087,mixed_local_spatial_buffering,M_tot,1.000000,STRICTLY_LOCAL,4.696968e-01,3.319892e-05,0.024824,0.475292,provisional_accepted_ensemble_tag_pending_step...
1,1_DH_1_CONTROL,DH,CONTROL,1_DH_1_CONTROL__hybrid_optuna_098,mixed_local_spatial_buffering,M_tot,1.000000,STRICTLY_LOCAL,4.060930e-01,5.965698e-05,0.351798,0.645323,provisional_accepted_ensemble_tag_pending_step...
2,1_DH_2_CONTROL,DH,CONTROL,1_DH_2_CONTROL__hybrid_optuna_066,low_recruitment_local_storage,M_tot,1.000000,STRICTLY_LOCAL,9.859677e-305,8.850469e-303,0.351407,0.570691,provisional_accepted_ensemble_tag_pending_step...
3,1_DH_2_CONTROL,DH,CONTROL,1_DH_2_CONTROL__hybrid_optuna_068,mixed_local_spatial_buffering,M_tot,1.000000,BALANCED_OR_WEAK,4.178607e-01,4.909472e-01,0.479434,0.518783,provisional_accepted_ensemble_tag_pending_step...
4,1_DH_2_CONTROL,DH,CONTROL,1_DH_2_CONTROL__hybrid_optuna_072,low_recruitment_local_storage,M_tot,1.000000,STRICTLY_LOCAL,9.859677e-305,0.000000e+00,0.166621,0.330441,provisional_accepted_ensemble_tag_pending_step...
5,1_DH_2_CONTROL,DH,CONTROL,1_DH_2_CONTROL__hybrid_optuna_098,mixed_local_spatial_buffering,M_tot,1.000000,STRICTLY_LOCAL,4.742352e-01,1.230573e-05,0.024763,0.476031,provisional_accepted_ensemble_tag_pending_step...
6,1_VH_1_CONTROL,VH,CONTROL,1_VH_1_CONTROL__hybrid_optuna_002,kir_dominant_local_buffering,M_tot,1.000000,STRICTLY_LOCAL,9.240382e-289,1.820215e-286,0.152815,0.843931,provisional_accepted_ensemble_tag_pending_step...
7,1_VH_1_CONTROL,VH,CONTROL,1_VH_1_CONTROL__hybrid_optuna_004,kir_dominant_local_buffering,M_tot,1.000000,STRICTLY_LOCAL,9.859677e-305,1.065506e-303,0.032881,0.965949,provisional_accepted_ensemble_tag_pending_step...
8,1_VH_1_CONTROL,VH,CONTROL,1_VH_1_CONTROL__hybrid_optuna_075,kir_dominant_local_buffering,M_tot,1.000000,STRICTLY_LOCAL,9.859677e-305,1.062399e-301,0.334101,0.664559,provisional_accepted_ensemble_tag_pending_step...
9,1_VH_1_CONTROL,VH,CONTROL,1_VH_1_CONTROL__hybrid_post_002,kir_dominant_local_buffering,M_tot,1.000000,STRICTLY_LOCAL,9.240382e-289,1.820215e-286,0.152815,0.843931,provisional_accepted_ensemble_tag_pending_step...


,file_id,region,condition,candidate_id,sweep,current_na,mode_M0,mode_M_decay,mode_M_rise,mode_M_tot,buffering_phenotype,phenotype_claim_scope
0,1_DH_1_CONTROL,DH,CONTROL,1_DH_1_CONTROL__hybrid_optuna_087,1,50,BALANCED_OR_WEAK,BALANCED_OR_WEAK,STRICTLY_LOCAL,STRICTLY_LOCAL,mixed_local_spatial_buffering,provisional_accepted_ensemble_tag_pending_step...
1,1_DH_1_CONTROL,DH,CONTROL,1_DH_1_CONTROL__hybrid_optuna_087,2,75,BALANCED_OR_WEAK,BALANCED_OR_WEAK,STRICTLY_LOCAL,STRICTLY_LOCAL,mixed_local_spatial_buffering,provisional_accepted_ensemble_tag_pending_step...
2,1_DH_1_CONTROL,DH,CONTROL,1_DH_1_CONTROL__hybrid_optuna_087,3,100,BALANCED_OR_WEAK,BALANCED_OR_WEAK,STRICTLY_LOCAL,STRICTLY_LOCAL,mixed_local_spatial_buffering,provisional_accepted_ensemble_tag_pending_step...
3,1_DH_1_CONTROL,DH,CONTROL,1_DH_1_CONTROL__hybrid_optuna_087,4,125,BALANCED_OR_WEAK,BALANCED_OR_WEAK,STRICTLY_LOCAL,STRICTLY_LOCAL,mixed_local_spatial_buffering,provisional_accepted_ensemble_tag_pending_step...
4,1_DH_1_CONTROL,DH,CONTROL,1_DH_1_CONTROL__hybrid_optuna_087,5,150,BALANCED_OR_WEAK,MIXED_SPATIAL,STRICTLY_LOCAL,STRICTLY_LOCAL,mixed_local_spatial_buffering,provisional_accepted_ensemble_tag_pending_step...
5,1_DH_1_CONTROL,DH,CONTROL,1_DH_1_CONTROL__hybrid_optuna_087,6,175,BALANCED_OR_WEAK,MIXED_SPATIAL,STRICTLY_LOCAL,STRICTLY_LOCAL,mixed_local_spatial_buffering,provisional_accepted_ensemble_tag_pending_step...
6,1_DH_1_CONTROL,DH,CONTROL,1_DH_1_CONTROL__hybrid_optuna_098,1,50,BALANCED_OR_WEAK,MIXED_LOCAL,STRICTLY_LOCAL,STRICTLY_LOCAL,mixed_local_spatial_buffering,provisional_accepted_ensemble_tag_pending_step...
7,1_DH_1_CONTROL,DH,CONTROL,1_DH_1_CONTROL__hybrid_optuna_098,2,75,BALANCED_OR_WEAK,MIXED_LOCAL,STRICTLY_LOCAL,STRICTLY_LOCAL,mixed_local_spatial_buffering,provisional_accepted_ensemble_tag_pending_step...
8,1_DH_1_CONTROL,DH,CONTROL,1_DH_1_CONTROL__hybrid_optuna_098,3,100,BALANCED_OR_WEAK,MIXED_LOCAL,STRICTLY_LOCAL,STRICTLY_LOCAL,mixed_local_spatial_buffering,provisional_accepted_ensemble_tag_pending_step...
9,1_DH_1_CONTROL,DH,CONTROL,1_DH_1_CONTROL__hybrid_optuna_098,4,125,BALANCED_OR_WEAK,MIXED_LOCAL,STRICTLY_LOCAL,STRICTLY_LOCAL,mixed_local_spatial_buffering,provisional_accepted_ensemble_tag_pending_step...


## 9. Phenotype counts by region, condition, and window

In [12]:
phenotype_counts = result["phenotype_counts_by_region_condition_window"]
phenotype_counts.sort_values(["region", "condition", "window", "buffering_phenotype"]).head(40)

,region,condition,window,buffering_phenotype,n_candidate_sweeps,n_candidates,n_cells
0,DH,CONTROL,M0,available_surface_voltage_coupled_but_ionic_re...,18,3,3
1,DH,CONTROL,M0,low_recruitment_local_storage,12,2,2
2,DH,CONTROL,M0,mixed_local_spatial_buffering,72,12,6
3,DH,CONTROL,M_decay,available_surface_voltage_coupled_but_ionic_re...,12,2,2
4,DH,CONTROL,M_decay,long_range_recruited_spatial_buffering,18,7,6
5,DH,CONTROL,M_decay,low_recruitment_local_storage,12,2,2
6,DH,CONTROL,M_decay,mixed_local_spatial_buffering,60,12,6
7,DH,CONTROL,M_rise,available_surface_voltage_coupled_but_ionic_re...,6,1,1
8,DH,CONTROL,M_rise,kir_dominant_local_buffering,6,2,2
9,DH,CONTROL,M_rise,low_recruitment_local_storage,13,3,2


## 10. Explicit claim-scope table

In [13]:
claims = result["claim_scope_table"]
claims[[
    "claim_topic", "allowed_pre_step06_claim", "forbidden_pre_step06_claim",
    "required_next_step", "n_clusters", "n_independent_cells", "n_candidates",
    "cluster_evidence_status", "proxy_validity_status",
]]


,claim_topic,allowed_pre_step06_claim,forbidden_pre_step06_claim,required_next_step,n_clusters,n_independent_cells,n_candidates,cluster_evidence_status,proxy_validity_status
0,mechanism_diversity,accepted_compensation_or_mixed_mechanism_struc...,candidate_degenerate_regimes,Step 06 predictive and perturbation validation,3,30,2098,candidate_regime_screen,proxy_not_failed_majority
1,region_enrichment,population_level_cell_association_not_phenotyp...,paired_pharmacology_or_animal_level_phenotype,larger stratified validation and Step 06 robus...,3,30,2098,candidate_regime_screen,proxy_not_failed_majority


In [14]:
assert (flux["simulation_status"] == "ok").any()
assert {"region", "condition", "mechanism_cluster", "cluster_evidence_status"}.issubset(clusters.columns)
assert "K_bath_middle_used" in flux.columns
assert flux.groupby("candidate_id")["K_bath_middle_used"].nunique().min() > 1
assert not windowed.empty
assert windowed["dKs_activation_score"].between(0.0, 1.0).all()
assert "buffering_phenotype" in phenotypes.columns
assert claims["forbidden_pre_step06_claim"].str.contains("candidate_degenerate_regimes").any()
assert not claims["allowed_pre_step06_claim"].str.contains("candidate_degenerate_regimes").any()
if clusters["cluster_evidence_status"].eq("insufficient_evidence").any():
    assert claims.loc[claims["claim_topic"] == "mechanism_diversity", "allowed_pre_step06_claim"].iloc[0] == "insufficient_evidence_for_candidate_mechanism_regimes"
print(f"Step 05 full-scope notebook validated and outputs saved under {out_dir}")


Step 05 full-scope notebook validated and outputs saved under /home/xav/code/astromodel_proving/outputs/mechanisms


## Post-execution scientific status

Executed status for reviewer response: Step 05 ran on the full 2098 accepted-candidate ensemble and produced 12,588 candidate-sweep rows, 50,272 windowed mechanism rows, 2098 phenotype-tag rows, and 3 candidate mechanism clusters across 30 independent cells. This supports R5 model characterization and a candidate-regime screen, including the reused local/spatial/recruitment vocabulary from the characterization notebook. The maturity level remains `candidate_regime_screen`: phenotype and mechanism labels are descriptive until Step 06 predictive/perturbation support and later guardrails mature.